File: agent/src/main/scala/io/scalac/mesmer/agent/akka/persistence/impl/RecoveryStartedAdvice.scala
package io.scalac.mesmer.agent.akka.persistence.impl

import java.lang.invoke.MethodHandle

import _root_.akka.actor.typed.scaladsl.ActorContext
import _root_.akka.persistence.typed.PersistenceId
import net.bytebuddy.asm.Advice._

import io.scalac.mesmer.core.event.EventBus
import io.scalac.mesmer.core.event.PersistenceEvent.RecoveryStarted
import io.scalac.mesmer.core.model._
import io.scalac.mesmer.core.util.ReflectionFieldUtils
import io.scalac.mesmer.core.util.Timestamp

object RecoveryStartedAdvice {

  @OnMethodEnter
  def enter(
    @Argument(0) context: ActorContext[_],
    @This self: AnyRef
  ): Unit = {

    println("recovery started agent")
    val path = context.self.path.toPath

    val replayingSnapshotsSetupGetter: MethodHandle =
      ReflectionFieldUtils.getGetter("akka.persistence.typed.internal.ReplayingSnapshot", "setup")

    val behaviorSetupPersistenceId: MethodHandle =
      ReflectionFieldUtils.getGetter("akka.persistence.typed.internal.BehaviorSetup", "persistenceId")

    val persistenceIdHandle: MethodHandle =
      ReflectionFieldUtils.chain(replayingSnapshotsSetupGetter, behaviorSetupPersistenceId)

    val persistenceId = persistenceIdHandle.invoke(self).asInstanceOf[PersistenceId]

    EventBus(context.system).publishEvent(RecoveryStarted(path, persistenceId.id, Timestamp.create()))
  }
}

File: otel-extension/src/main/java/akka/persistence/RecoveryStartedAdvice.java
package akka.persistence;

import akka.actor.typed.scaladsl.ActorContext;
import io.scalac.mesmer.otelextension.instrumentations.akka.persistence.impl.RecoveryCompletedImpl;
import io.scalac.mesmer.otelextension.instrumentations.akka.persistence.impl.RecoveryStartedImpl;
import net.bytebuddy.asm.Advice;

public class RecoveryStartedAdvice {


    @Advice.OnMethodEnter
    public static void onStarted(@Advice.Argument(0) ActorContext<?> context, @Advice.This Object self) {
        System.out.println("STARTED FIRST");
        RecoveryStartedImpl.enter(context, self);
    }

    @Advice.OnMethodExit(onThrowable = Throwable.class)
    public static void onExit(@Advice.Thrown(readOnly = false) Throwable error) {
        if(error != null) {
            System.out.println("ERROR OCCURRED " + error.getMessage());
            error = null;
        } else {
            System.out.println("FINISHED SUCCESSFULLY STARTING RECOVERY");
        }
    }
}
